### DISCUSSION TITLE
EL Taglib implementation/other remaining classes from old Struts library

### DISCUSSION THREAD

**Discussion by lukamalovic** [ste-gr/struts1-reloaded#18] [General]
@ste-gr 

Hi Stefan/all those concerned,

I wanted to ask whether there were any plans on implementing the remaining several classes from the previous Struts library: namely, providing full implementations for the strutsel.taglib.* classes as well as adding their corresponding BeanInfo.java files.

Thanks,
Luka Malovic

**Comment by ste-gr** [comment]
@lukamalovic 

Hi Luka,

the tags themselves should be complete (like Struts 1.3.10). With commit 5917b85 I removed the BeanInfo.java-files. As I remember correctly, I removed them because I thought they were not necessary.
If I was wrong, I can restore them.

Greetings
Stefan

**Reply by TimCincotta** [reply]
Hello Stefan,

Thank you for all your efforts on this project.  Regarding the removal of the BeanInfo.java files, my guess is that we will need to have them added back.  The application we support, which we will be utilizing this new Struts implementation for, has a number of custom tags which extend the EL...Tag classes.  Though I haven't yet tested this fully, the javadoc for the EL...TagBeanInfo explains the need for these classes:
```
/**
 * This is the <code>BeanInfo</code> descriptor for the
 * <code>org.apache.strutsel.taglib.html.ELRadioTag</code> class.  It is
 * needed to override the default mapping of custom tag attribute names to
 * class attribute names. <p> This is because the value of the unevaluated EL
 * expression has to be kept separately from the evaluated value, which is
 * stored in the base class. This is related to the fact that the JSP compiler
 * can choose to reuse different tag instances if they received the same
 * original attribute values, and the JSP compiler can choose to not re-call
 * the setter methods, because it can assume the same values are already set.
 */
```
I have needed to recreate the EvalHelper class in my local instance of struts1, as well - we make extensive use of that class.  (I, of course, needed to change imports to jakarta.* references, and add the jakarta.el:jakarta.el-api dependency in the struts-el/pom.xml file)

Requesting your consideration of restoring these classes.  If you would like an example of how our codebase is using them, please let me know. 

Kind regards,
Tim

**Comment by ste-gr** [comment]
@lukamalovic, @TimCincotta

Hallo Luka and Tim,

I have now I taken the time to revisit my changes. The changes was necessary because the upgrade from JSP 2.0 to 2.1. With JSP 2.1, it is much easier to use JSP-EL than before.

**JSP 2.0 and before:**
EL-Tag-class extends Tag-class:
* With `BeanInfo` the tag-attributes are stored 1:1 to `EL-Attributes`.
* The `EL-Attributes` are evaluated with the `EvalHelper`-class.
* The evaluated attributes are stored to the `Attributes` of the Tag-class.
* The Tag-class can render the HTML-code.

**JSP 2.1+:**
The EL-Tag-class only extends the Tag-class to differentiate between the two classes in the `tld`-file:
* JSP-Library takes the attributes, evaluates them and then stores them in the `Attributes` of the Tag-class.
* The Tag-class can render the HTML-code.

Therefore, the EL-Tag-class no longer has any content and the `BeanInfo` is no longer necessary.

I think this is also the reason why the libraries `el` and `faces` were marked as `dormant` by Apache.

Maybe I can help you to update your codebase using the example of a tag.

Greetings
Stefan

**Comment by TimCincotta** [comment]
Hello Stefan-

I appreciate your help in looking into this - that explanation helps me understand things better.

I will have a discussion with Luka in light of what you have explained, and we can try refactoring things accordingly.  Indeed, if we run into issues, we will take you up on your offer.

Regards,
Tim
________________________________

**Reply by ste-gr** [reply]
Hi Tim,

I'm glad if I can help you. Perhaps if your changes work, you could write a short upgrade documentation to help other people as well.

Thank you and
regards,
Stefan

**Comment by ste-gr** [comment]
@lukamalovic, @TimCincotta

Hello Luka and Tim,

how are things going with you? Do you need any help? Are there any changes needed in `Struts1`?
If no changes are needed in `Struts1`, I would like to publish this version.

Regards
Stefan
